Annotations main view

mbarto opened the pull request:
## Description
This pull request completes the annotations functionalities, adding the main panel with annotations list / detail / editing.

## Issues
 - Fix #2157

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [x] Feature

**What is the current behavior?** (You can also link to an open issue here)
The annotations panel is not open when clicking on the related burger menu option

**What is the new behavior?**
The panel is open and working as expected.

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

offtherailz commented on web/client/epics/annotations.js:
```diff
@@ -58,34 +59,38 @@ const mergeGeometry = (features) => {
     addAnnotationsLayerEpic: (action$, store) =>
     action$.ofType(MAP_CONFIG_LOADED)
         .switchMap(() => {
             const annotationsLayer = head(store.getState().layers.flat.filter(l => l.id === 'annotations'));
```
use layers selector and then filter

offtherailz commented on web/client/plugins/Annotations.jsx:
```diff
@@ -10,53 +10,41 @@ const React = require('react');
+    config: state.annotations && state.annotations.config,
     editing: state.annotations && state.annotations.editing,
     drawing: state.annotations && !!state.annotations.drawing,
     styling: state.annotations && !!state.annotations.styling,
```
All state slices are replicated in many places. You could simply use createSelector

offtherailz commented on web/client/plugins/Annotations.jsx:
```diff
@@ -74,7 +62,112 @@ const AnnotationsInfoViewer = connect((state) => ({
+    annotations: state.layers && state.layers.flat && head(state.layers.flat.filter(l => l.id === 'annotations')) && head(state.layers.flat.filter(l => l.id === 'annotations')).features || [],
+    current: state.annotations && state.annotations.current || null,
+    editing: state.annotations && state.annotations.editing,
+    filter: state.annotations && state.annotations.filter || ''
```
same as above,  some of thees should go in a selector

offtherailz commented on web/client/plugins/Annotations.jsx:
```diff
@@ -10,53 +10,41 @@ const React = require('react');
-})(require('../components/mapcontrols/annotations/Annotations'));
+
+const AnnotationsEditor = connect((state) => ({
+    config: state.annotations && state.annotations.config,
```
use selector

offtherailz commented:
Some notes about usage:
 - While editing a new, if I close the annotation panel, it should ask confirm before closing. 
 - If I start drawing points and I click on thrash button, It seems like I'm able to create points (the tool is active and the map shows the same tooltip) but I can't. 
 - It is not clear how can I stop adding point. I propose to allow adding only one point each time and disable button when added. If a user want to add another point, he have to click again on the button.
 - I can not edit title or description when I clicked first on "add point". Maybe on click on the annotation panel you could disable editing
 - Is difficoult to exit from the edit mode, maybe we should allow to exit by clicking on the panel and/or hitting "ESC" (or right click). 
- With disabled featureinfo, hovering the mouse on annotation, I see I can click but I can't.
- Buttons are too near to the bottom of the page, so they are not too visible
- Errors should be more visible (e.g. we should use red color)

mbarto commented:
Done the following:
 - Added confirm before closing in editing mode (both on panel and identify)
 - restored correct drawing mode after geometry delete
 - set single point by default (multiGeometry enabled with configuration)
 - disabled edting fields during geometry drawing
- removed cursor pointer css from annotations markers when not in identify mode
- moved buttons on the top of the page
- changed css of errors to show them in red
- removed annotations layer when there are no more annotations on the map

About the following, this is my opinion:
 - Annotations layer should be always on top? NO
 - Does it make sense to show it in TOC? YES, because you may want to show/hide annotations

simboss commented:
I am happy to have annotations as a separate layer in the TOC.

In the future we might have the concept of per layer annotations.

offtherailz commented:
Good, it makes sense.
I'm going to review latest changes

offtherailz commented:
I found a regression with editing of multigeom. Try to edit 1 geome of an existing multigeom and save. The modified disappears. I checked and it is working on current stable.
